types/gl-react/tslint.json

```diff
@@ -0,0 +1,6 @@
+{
+	"extends": "dtslint/dt.json",
+  "rules": {
+    "no-unnecessary-class": false
```

Comment:
This doesn't strictly seem necessary (and declaring things like `Shaders` as classes is also inaccurate). Either switch your classes to `namespace`s, or objects like

```ts
const Shaders: {
    create(shaders: { [key: string]: ShaderDefinition }): ShadersSheet;
    // ...
};
```

Reply:
I tried with different combinations, but now I managed to get this the right way.

So the three (earlier) classes with statics are now written as:

```
export namespace Xxx {
  type yyy = () => void;
  ...
}
```

Maybe there could be such examples somewhere documented. Many Open Source projects use generic JavaScript objects, and there's many ways to write types for them (one that I tried was `const foo = { bar: () => {} }; export type foo = typeof foo;` which was not good since it did lost the context of class/object/interface names.

Reply:
Actually namespace doesn't work, since `gl-react` uses `Shaders.create( ... )` interface, which doesn't work. So that will limit the possibilities.

Reply:
And with `Uniform` class having `Node | Bus` type, it's even not possible to write it as const. So I'm reverting back to classes and `no-unnecessary-class: false`.

Reply:
Here's the code that I tried:
```
export namespace Shaders {
  type create = (shaders: { [key: string]: ShaderDefinition }) => ShadersSheet;
  type getName = (shaderIdentifier: ShaderIdentifier) => string;
  type getShortName = (shaderIdentifier: ShaderIdentifier) => string;
  type get = (shaderIdentifier: ShaderIdentifier) => ShaderInfo;
}

export namespace Uniform {
  type Backbuffer = string;
  type backbufferFrom = (node: Node | Bus) => { type: string, node: Node | Bus };
  type Resolution = string;
  type textureSize = (obj: any) => { type: string, obj: any };
  type textureSizeRatio = (obj: any) => { type: string, obj: any, ratio: boolean };
}

export namespace Visitors {
  type add = (visitor: Visitor) => void;
  type remove = (visitor: Visitor) => void;
  type get = () => Visitor[];
}
```

If I try writing the namespaces/classes as consts, it would be like:
```
const Shaders = {
 create: (shaders: { [key: string]: ShaderDefinition }) => ({ '0': { type: '', id: '' } }),
 getName: (shaderIdentifier: ShaderIdentifier) => '',
 getShortName: (shaderIdentifier: ShaderIdentifier) => '',
 get: (shaderIdentifier: ShaderIdentifier) => ({ type: '', id: '' })
};
export type Shaders = typeof Shaders;
```

This might seem to work for Shaders (except my editor's TypeScript plugin giving a lot of "errors"), but it doesn't work for `Uniform` as `backbufferFrom` response might contain `Node` or `Bus` as `node`.

Reply:
Maybe I'm missing something, but if you want a type for each of your constants, why doesn't the following work?

```ts
export interface Uniform {
  Backbuffer: string;
  backbufferFrom(node: Node | Bus): { type: string, node: Node | Bus };
  Resolution: string;
  textureSize(obj: any): { type: string, obj: any };
  textureSizeRatio(obj: any): { type: string, obj: any, ratio: boolean };
}

export const Uniform: Uniform;
```

Reply:
I tried also an interface. When using `Shaders.create( ... )` it gives an error cannot use type as value. I can get the exact error Tomorrow. My first try was with inteface actually. I can also share the example code also.

I know, it would be really nice to get rid of the exception. And in `gl-react` these ”classes with only statics) are generic objects with properties which have anonymous functions. I don’t know if that coding is a good style, but it’s very common in non-TypeScript-style ES2015/2016/2017 code.

Reply:
>  And in gl-react these ”classes with only statics) are generic objects with properties which have anonymous functions. I don’t know if that coding is a good style, but it’s very common in non-TypeScript-style ES2015/2016/2017 code.

It's perfectly acceptable, and that's really why I'm trying to avoid writing these as classes - they weren't defined as classes in the first place so it's inaccurate to declare them as so. So just trying to make sure we don't accidentally let users `new` one of them up 😄

Reply:
I did manage to get the example code to work by using a namespace, and const/function inside it. It doesn't match the original either, but goes with tslint defaults so let it be like that. @DanielRosenwasser, please confirm the changes are OK now :-) ...